### PR TITLE
Refactor inferschematype.d.ts

### PR DESCRIPTION
**Pull request by mohammad0-0ahmad**
Related to https://github.com/Automattic/mongoose/pull/12007

### CODE REVIEW

File: types/inferschematype.d.ts [hunk 1]
```diff
@@ -76,31 +78,20 @@ type RequiredPathBaseType = { required: true | [true, string | undefined] };
  */
 type PathWithTypePropertyBaseType<TypeKey extends TypeKeyBaseType> = { [k in TypeKey]: any };
```
Comment (iammola) [inline types/inferschematype.d.ts:80]:
```suggestion

/**
 * @summary Array type with undefined default
 * @description It helps to check if a path is an array and its default is undefined
 * @param {TypeKey} TypeKey A generic of literal string type."Refers to the property used for path type definition".
 */
type UndefinedDefaultArray<TypeKey extends TypeKeyBaseType> = Record<TypeKey, ArrayConstructor | any[]> & { default: undefined };
```

File: types/inferschematype.d.ts [hunk 2]
```diff
@@ -76,31 +78,20 @@ type RequiredPathBaseType = { required: true | [true, string | undefined] };
...
  * @returns required paths keys of document definition.
  */
 type RequiredPathKeys<T, TypeKey extends TypeKeyBaseType> = {
-  [K in keyof T]: T[K] extends RequiredPathBaseType ? IfEquals<T[K], any, never, K>
-    // Check if the path's type is an array
-    : PathTypeDefinition<T[K], TypeKey> extends any[]
-      // If the path is an array and is defaulted to undefined, then we remove it from the required keys
-      ? T[K] extends { default: undefined } ? never : K
-      : never;
+  [K in keyof T]: T[K] extends RequiredPathBaseType<TypeKey> ? IfEquals<T[K], any, never, K> : never;
```
Comment (iammola) [inline types/inferschematype.d.ts:88]:
```suggestion
  [K in keyof T]: T[K] extends RequiredPathBaseType<TypeKey> ? IfEquals<T[K], any, never, T[K] extends UndefinedDefaultArray<TypeKey> ? never : K> : never;
```

File: types/inferschematype.d.ts [hunk 3]
```diff
@@ -110,20 +101,17 @@ type RequiredPaths<T, TypeKey extends TypeKeyBaseType> = {
...
  * @returns optional paths keys of document definition.
  */
 type OptionalPathKeys<T, TypeKey extends TypeKeyBaseType> = {
-  [K in keyof T]: T[K] extends RequiredPathBaseType ? never
-    // Check if the path's type is an array
-    : PathTypeDefinition<T[K], TypeKey> extends any[]
-      // If the path is an array and is defaulted to undefined, then we add it to the optional keys
-      ? T[K] extends { default: undefined } ? K : never
-      : K;
+  [K in keyof T]: T[K] extends RequiredPathBaseType<TypeKey> ? never : K;
```
Comment (iammola) [inline types/inferschematype.d.ts:108]:
```suggestion
  [K in keyof T]: T[K] extends RequiredPathBaseType<TypeKey> ? T[K] extends UndefinedDefaultArray<TypeKey> ? K : never : K;

```

File: types/inferschematype.d.ts (hunk 1 above)
Comment (iammola) [reply types/inferschematype.d.ts:80]: You'll probably want to change this helper type name and description to something else.

**Review by iammola — COMMENTED** [review]
Hi @mohammad0-0ahmad 
Your code is perfectly what was needed.

The changes I've requested are so array paths with `{ default: undefined }` are also marked as optional. If not all arrays even the ones that have been explicitly marked to default to undefined are marked as required.

Is that something you're fine with? If so. I'll merge your PR and discard the changes. Thanks for keeping up with my troubles 🙏.

**Comment by mohammad0-0ahmad** [comment]
> Hi @mohammad0-0ahmad Your code is perfectly what was needed.
> 
> The changes I've requested are so array paths with `{ default: undefined }` are also marked as optional. If not all arrays even the ones that have been explicitly marked to default to undefined are marked as required.
> 
> Is that something you're fine with? If so. I'll merge your PR and discard the changes. Thanks for keeping up with my troubles 🙏.

Good job @iammola,  I forgot to cover that.
I've pushed a new commit that should fix it.
Please feel free to suggest any improvements.

**Comment by iammola** [comment]
Yes @mohammad0-0ahmad, the latest commit is much cleaner and it works perfectly 👏

Thank you 🙏

**Pull request merged**